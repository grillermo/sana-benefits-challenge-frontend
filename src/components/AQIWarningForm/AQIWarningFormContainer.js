import { useEffect, useState, useContext } from 'react'
import { useLocation } from 'wouter'
import { isEmpty } from 'lodash'

import AuthContext from '../../contextProviders/authContext'

import AQIWarningForm from './AQIWarningForm'
import AirQualityWidget from '../AirQualityWidget'
import { getAQIWarning, getAQI, saveAQIWarning } from '../../utils/apiClient'

export function handleSubmit(setSubmitting, aqiWarning) {
  setSubmitting(true)

  return saveAQIWarning(aqiWarning).then(() => (setSubmitting(false)))
                                   .catch(() => (setSubmitting(false)))
}

export function handleSuggestionChange(suggestion, setAqiWarning, aqiWarning) {
  const _aqiWarning = {
    ...aqiWarning,
    location: suggestion.label,
    latitude: suggestion.location.lat,
    longitude: suggestion.location.lng,
  }

  setAqiWarning(_aqiWarning)
}

export function handleThresholdChange(event, setAqiWarning, aqiWarning) {
  let value

  if(isEmpty(event.target.value)) {
    value = 0
  } else {
    value =  parseInt(event.target.value)
  }

  const _aqiWarning = {
    ...aqiWarning,
    threshold: value
  }

  setAqiWarning(_aqiWarning)
}


export default function AQIWarningFormContainer() {
  const { state: authState }        = useContext(AuthContext);
  const [aqiWarning, setAqiWarning] = useState({ location: '', threshold: 0 })
  const [aqi, setAqi]               = useState(undefined)
  const [location, setLocation]     = useLocation();

  // Initial load with a possible saved aqiWarning
  useEffect(() => {
    getAQIWarning().then((data) => {
      // The user does not have saved aqiWarning
      if(!data || !data.id) {
        return
      }

      const _aqiWarning = {
        ...aqiWarning,
        location: data.location,
        latitude: data.latitude,
        longitude: data.longitude,
        threshold: data.threshold
      }

      setAqiWarning(_aqiWarning)
    })
  },[])
  
  // On latitude longitude updates
  useEffect(() => {
    getAQI(aqiWarning.latitude, aqiWarning.longitude).then((aqi) => (setAqi(aqi)))
  },[aqiWarning.latitude, aqiWarning.longitude])
  
  useEffect(() => {
    if(!authState.isAuthenticated) {
      return setLocation('/sign_in')
    }
  }, [authState]);

  return (
    <div className='AQIWarningContainer'>
      <AQIWarningForm
        handleSuggestionChange={(suggestion) => (
          handleSuggestionChange(suggestion, setAqiWarning, aqiWarning)
        )}
        handleThresholdChange={(event) => (
          handleThresholdChange(event, setAqiWarning, aqiWarning)
        )}
        aqiWarning={aqiWarning}
        handleSubmit={(values, { setSubmitting }) => (
          handleSubmit(setSubmitting, aqiWarning)
        )}
      />

      <AirQualityWidget 
        currentLevel={aqi}
        threshold={aqiWarning.threshold}
      />
    </div>
  )
}
