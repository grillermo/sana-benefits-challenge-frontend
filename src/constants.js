export const API_URL = process.env.REACT_APP_API_URL
export const AQI_TOKEN = process.env.REACT_APP_AQI_TOKEN
