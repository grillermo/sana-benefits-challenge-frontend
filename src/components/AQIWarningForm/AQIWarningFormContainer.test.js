import { mount, shallow } from 'enzyme'
import { act, render, screen } from '@testing-library/react'

import googleStub from './Geosuggest/GoogleSutb'
import axios from 'axios'

import AQIWarningFormContainer, {
  handleSubmit,
  handleSuggestionChange,
  handleThresholdChange
} from './AQIWarningFormContainer'
import AuthContext from '../../contextProviders/authContext'

import { API_URL } from '../../constants'

jest.mock('axios')

function mockPOSTAxiosCall(saveAQIWarningMock) {
  return axios.post.mockImplementationOnce(() => Promise.resolve(saveAQIWarningMock))
}

function mockGETAxiosCalls(getAQIWarningMock, getAQIMock, saveAQIWarningMock) {
  const defaultGetAQIWarning = {
    data: {
      latitude: 0,
      longitude: 0,
      location: '',
      threshold: 0
    }
  }
  const defaultgetAQIMock = {
    data: {
      data: {
        aqi: 0
      }
    }
  }
  const defaultSaveAQIWarningMock = {
    data: {
      data: {
        aqi: 0
      }
    }
  }

  return axios.get.mockImplementation((url) => {
    if(url === 'aqi_warning') {
      // Mock getAQIWarning
      return Promise.resolve(getAQIWarningMock || defaultGetAQIWarning)
    } else if (url.startsWith('https://api.waqi.info/feed/geo')) {
      // Mock getAQI
      return Promise.resolve(getAQIMock || defaultgetAQIMock)
    }
  })
}


describe('AQIWarningFormContainer', () => {
  let getAQIWarningMock, getAQIMock
  let isAuthenticated = true

  async function renderAQIWarningFormContainer() {
    return render(
      <AuthContext.Provider
        value={{state: { isAuthenticated: isAuthenticated }}}
      >
        <AQIWarningFormContainer />
      </AuthContext.Provider>
    )
  }

  beforeEach(() => {
    window.google = googleStub()
  })

  describe('an unauthenticated user', () => {
    isAuthenticated = false

    beforeEach(() => {
      mockGETAxiosCalls()
    })

    it('redirecs to the sign_in path', async () => {
      await act(() => renderAQIWarningFormContainer() )

      expect(location.pathname).toBe('/sign_in')
    })
  })

  describe('a previously created aqiWarning exists', () => {
    beforeEach(() => {
      getAQIWarningMock = {
        data: {
          id: 1,
          latitude: '19.002529441654687',
          longitude: '-98.26751911537335',
          location: 'Texas',
          threshold: 70
        }
      }

      mockGETAxiosCalls(getAQIWarningMock)
    })

    it('sets the aqiWarning values on the form', async () => {
      await act(() => renderAQIWarningFormContainer() )

      expect(screen.getByLabelText('Your location:').value).toBe(getAQIWarningMock.data.location)
      expect(screen.getByLabelText('Your warning threshold:').value).toBe(
        getAQIWarningMock.data.threshold.toString()
      )
    })

    it('matches the snapshot', async () => {
      let page
      await act(async () => {
        page = await renderAQIWarningFormContainer()
      })

      expect(page).toMatchSnapshot();
    })
  })

  describe('the user does not have an aqiWarning', () => {
    beforeEach(() => {
      getAQIWarningMock = {
        data: { }
      }

      mockGETAxiosCalls(getAQIWarningMock)
    })

    it('sets default values on the form', () => {
    })

    it('matches the snapshot', async () => {
      let page
      await act(async () => {
        page = await renderAQIWarningFormContainer()
      })

      expect(page).toMatchSnapshot();
    })
  })

  describe('handleSubmit', () => {
    let setSubmitting
    const jwt = 'eyJhbGciOiJIUzI1NiJ9'
    let AQIWarningMock 

    beforeEach(() => {
      AQIWarningMock = {
        latitude: '19.002529441654687',
        longitude: '-98.26751911537335',
        location: 'Texas',
        threshold: 70
      }

      mockPOSTAxiosCall(AQIWarningMock)

      // The token is needed for signed in calls
      window.localStorage.setItem('token', jwt)

      setSubmitting = jest.fn();
    })

    it('calls saveAQIWarning', async () => {
      const response = await handleSubmit(setSubmitting, AQIWarningMock)

      expect(axios.post).toHaveBeenCalledWith(
        'aqi_warning',
        {
          aqi_warning: AQIWarningMock
        },
        {
          baseURL: API_URL,
          timeout: 1000,
          headers: {
            'Authorization': 'Bearer '+jwt
          }
        }
      )
    })

    it('sets the submitting flag', async () => {
      const response = await handleSubmit(setSubmitting, AQIWarningMock)

      expect(setSubmitting).toHaveBeenCalledWith(true)
    })
  })

  describe('handleSuggestionChange', () => {
    it ('updates the aqiWarning location, latitude and longitude', () => {
      const suggestion = {
        label: 'Texas',
        location: {
          lat: '19.002529441654687',
          lng: '-98.26751911537335',
        }
      }
      const setAqiWarning = jest.fn()
      const aqiWarning = {
        threshold: 60,
      }
      
      handleSuggestionChange(suggestion, setAqiWarning, aqiWarning)

      expect(setAqiWarning).toHaveBeenCalledWith({
        ...aqiWarning,
        location: suggestion.label,
        latitude: suggestion.location.lat,
        longitude: suggestion.location.lng,
      })
    })
  })

  describe('handleThresholdChange', () => {
    let event, threshold
    const setAqiWarning = jest.fn()
    const aqiWarning = {}

    describe('with a numeric threshold value', () => {
      beforeEach(() => {
        threshold = 10
        event = { target: { value: threshold.toString() } }
      })

      it('sets the threshold on the aqiWarning', async () => {
        handleThresholdChange(event, setAqiWarning, aqiWarning)

        expect(setAqiWarning).toHaveBeenCalledWith({
          ...aqiWarning,
          threshold: threshold,
        })
      })
    })

    describe('with an empty threshold value', () => {
      beforeEach(() => {
        threshold = ''
        event = { target: { value: threshold.toString() } }
      })

      it('sets the threshold as 0 on the aqiWarning', async () => {
        handleThresholdChange(event, setAqiWarning, aqiWarning)

        expect(setAqiWarning).toHaveBeenCalledWith({
          ...aqiWarning,
          threshold: 0,
        })
      })
    })
  })
})
