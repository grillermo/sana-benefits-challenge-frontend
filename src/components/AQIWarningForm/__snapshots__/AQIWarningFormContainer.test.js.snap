// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AQIWarningFormContainer a previously created aqiWarning exists matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="AQIWarningContainer"
      >
        <form>
          <label>
            Your location:
            <div
              class="geosuggest"
            >
              <div
                class="geosuggest__input-wrapper"
              >
                <input
                  aria-expanded="false"
                  aria-owns="geosuggest__list"
                  autocomplete="nope"
                  class="geosuggest__input"
                  id="geosuggest__input"
                  name="location"
                  placeholder="Beverly hills"
                  role="combobox"
                  tabindex="0"
                  type="text"
                  value="Texas"
                />
              </div>
              <div
                class="geosuggest__suggests-wrapper"
              >
                <ul
                  class="geosuggest__suggests geosuggest__suggests--hidden"
                  id="geosuggest__list"
                  role="listbox"
                />
              </div>
            </div>
          </label>
          <label>
            Your warning threshold:
            <input
              name="threshold"
              placeholder="100"
              type="number"
              value="70"
            />
          </label>
          <button
            type="submit"
          >
            Save
          </button>
        </form>
        <div
          class="airQualityWidget danger"
        >
          <h1>
            0
          </h1>
          <strong>
            current level
          </strong>
          <p>
            Dangerous
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="AQIWarningContainer"
    >
      <form>
        <label>
          Your location:
          <div
            class="geosuggest"
          >
            <div
              class="geosuggest__input-wrapper"
            >
              <input
                aria-expanded="false"
                aria-owns="geosuggest__list"
                autocomplete="nope"
                class="geosuggest__input"
                id="geosuggest__input"
                name="location"
                placeholder="Beverly hills"
                role="combobox"
                tabindex="0"
                type="text"
                value="Texas"
              />
            </div>
            <div
              class="geosuggest__suggests-wrapper"
            >
              <ul
                class="geosuggest__suggests geosuggest__suggests--hidden"
                id="geosuggest__list"
                role="listbox"
              />
            </div>
          </div>
        </label>
        <label>
          Your warning threshold:
          <input
            name="threshold"
            placeholder="100"
            type="number"
            value="70"
          />
        </label>
        <button
          type="submit"
        >
          Save
        </button>
      </form>
      <div
        class="airQualityWidget danger"
      >
        <h1>
          0
        </h1>
        <strong>
          current level
        </strong>
        <p>
          Dangerous
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`AQIWarningFormContainer the user does not have an aqiWarning matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="AQIWarningContainer"
      >
        <form>
          <label>
            Your location:
            <div
              class="geosuggest"
            >
              <div
                class="geosuggest__input-wrapper"
              >
                <input
                  aria-expanded="false"
                  aria-owns="geosuggest__list"
                  autocomplete="nope"
                  class="geosuggest__input"
                  id="geosuggest__input"
                  name="location"
                  placeholder="Beverly hills"
                  role="combobox"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="geosuggest__suggests-wrapper"
              >
                <ul
                  class="geosuggest__suggests geosuggest__suggests--hidden"
                  id="geosuggest__list"
                  role="listbox"
                />
              </div>
            </div>
          </label>
          <label>
            Your warning threshold:
            <input
              name="threshold"
              placeholder="100"
              type="number"
              value="0"
            />
          </label>
          <button
            type="submit"
          >
            Save
          </button>
        </form>
        <div
          class="airQualityWidget danger"
        >
          <h1>
            0
          </h1>
          <strong>
            current level
          </strong>
          <p>
            Dangerous
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="AQIWarningContainer"
    >
      <form>
        <label>
          Your location:
          <div
            class="geosuggest"
          >
            <div
              class="geosuggest__input-wrapper"
            >
              <input
                aria-expanded="false"
                aria-owns="geosuggest__list"
                autocomplete="nope"
                class="geosuggest__input"
                id="geosuggest__input"
                name="location"
                placeholder="Beverly hills"
                role="combobox"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
            <div
              class="geosuggest__suggests-wrapper"
            >
              <ul
                class="geosuggest__suggests geosuggest__suggests--hidden"
                id="geosuggest__list"
                role="listbox"
              />
            </div>
          </div>
        </label>
        <label>
          Your warning threshold:
          <input
            name="threshold"
            placeholder="100"
            type="number"
            value="0"
          />
        </label>
        <button
          type="submit"
        >
          Save
        </button>
      </form>
      <div
        class="airQualityWidget danger"
      >
        <h1>
          0
        </h1>
        <strong>
          current level
        </strong>
        <p>
          Dangerous
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
